### PR TITLE
Add data transformation to output data in the script

### PR DESCRIPTION
**Pull request by r0x0d**
Data transformation is added in this commit to manipulate the convert2rhel assessment output to conform with the expected output by Insights

### CODE REVIEW

File: scripts/preconversion_assessment_script.py [hunk 1]
```diff
@@ -226,29 +373,28 @@ def main():
...
         print("Collecting status.")
-        combined_levels = collect_report_level(action_results=data["action_results"])
-        output.status = combined_levels[0]
+        combined_levels = collect_report_level(action_results=data["actions"])
+
         # Set the first position of the list as being the final status, that's
         # needed because `collect_report_level` will sort out the list with the
         # highest priority first.
-
+        output.status = combined_levels[0]
```
Comment (andywaltlova) [inline scripts/preconversion_assessment_script.py:382]: thinking if the function shouldn't return the highest one already, do we need the rest? if the report_level = the highest one, then I would just return the level

File: scripts/preconversion_assessment_script.py [hunk 2]
```diff
@@ -197,27 +258,113 @@ def cleanup(required_files=None):
...
+    detail_block = {
+        "remediation": {"context": ""},
+        "diagnosis": {"context": ""},
+    }
+
+    detail_block["remediation"]["context"] = new_message["remediation"]
+    detail_block["diagnosis"]["context"] = new_message["diagnosis"]
+
+    del new_message["remediation"]
+    del new_message["diagnosis"]
```
Comment (andywaltlova) [inline scripts/preconversion_assessment_script.py:303]: the properties always exist in original data? I mean .. are we safe from key errors here? maybe pop with default values would be safer

**Review by andywaltlova — APPROVED** [review]
Just few questions here and there, overall great changes! :+1:

File: scripts/preconversion_assessment_script.py [hunk 3]
```diff
@@ -154,34 +156,93 @@ def setup_convert2rhel(required_files=None):
...
+
 def install_convert2rhel():
-    subprocess.check_output(["yum", "install", "convert2rhel", "-y"])
-    subprocess.check_output(["yum", "update", "convert2rhel", "-y"])
+    """Install the convert2rhel tool to the system."""
+    output, returncode = run_subprocess(
+        ["yum", "install", "convert2rhel", "-y"],
+    )
+    if returncode:
+        print(
```
[the remark this answers is not in the document]
Comment (r0x0d) [reply scripts/preconversion_assessment_script.py:203]: Yeah, because that will be collected by the stdout of the worker. So, printing to the stdout is okay for now. 

We can take a look at a logging solution later to make it more robust, but for now, a simple print statement would be enough.

File: scripts/preconversion_assessment_script.py (hunk 2 above)
Comment (r0x0d) [reply scripts/preconversion_assessment_script.py:305]: That's a very good point. I have changed it to be `new_message.pop("remediation", "")` and `new_message.pop("diagnosis", "")`.

File: scripts/preconversion_assessment_script.py [hunk 4]
```diff
@@ -197,27 +258,113 @@ def cleanup(required_files=None):
...
+    format.
+    """
+    new_data = []
+    for action_id, result in raw_data["actions"].items():
+        # Format the results as a single list
+        for message in result["messages"]:
+            new_data.append(apply_message_transform(message, action_id))
+
+        new_data.append(apply_message_transform(result["result"], action_id))
+    return new_data
```
[the remark this answers is not in the document]
Comment (r0x0d) [reply scripts/preconversion_assessment_script.py:344]: So, I would not expect that the messages and actions are empty, as convert2rhel will always generate that even if they are just an empty dict/list. I can add a new test to check that even if we have empty dict/list we will not error out in this step. 

But one thing that I'm thinking right now is adding a new check to main.py to verify that we do have the json file to work with, otherwise, we just raise a ProcessError and exit the script.

[the remark this answers is not in the document]
Comment (andywaltlova) [reply scripts/preconversion_assessment_script.py:344]: :thinking: don't know what you mean, the json report from analysis is present only after the command is run? If for some reason the json report file wouldn't be created then the exception message could be enough

[the remark this answers is not in the document]
Comment (r0x0d) [reply scripts/preconversion_assessment_script.py:344]: Yeah, it will be basically checking if the file/contents exists. If not, we should bail out.

File: scripts/preconversion_assessment_script.py [hunk 5]
```diff
@@ -197,27 +258,115 @@ def cleanup(required_files=None):
         )
 
 
-def verify_required_files_are_present(required_files=None):
-    if required_files is None:
-        required_files = REQUIRED_FILES
-
+def verify_required_files_are_present(required_files):
+    """Verify if the required files are already present on the system."""
+    print("Checking if required files are present on the system.")
```
[the remark this answers is not in the document]
Comment (r0x0d) [reply scripts/preconversion_assessment_script.py:263]: I moved it the other way around... All prints are in the functions itself rather than the main one. 

I think that having it like that makes it more clean and encapsulates the message to be inside each function.

**Comment by r0x0d** [comment]
Codecov is failing to upload the results, but tests are green.

File: scripts/preconversion_assessment_script.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (andywaltlova) [reply scripts/preconversion_assessment_script.py:283]: okay, looks great in the main now!

File: scripts/preconversion_assessment_script.py [hunk 6]
```diff
@@ -11,6 +12,23 @@
     from urllib2 import urlopen
```
Comment (andywaltlova) [inline scripts/preconversion_assessment_script.py:12]: Only remaining question is for python3 usage, just for fun I tried to run it with python 3.6, and the script fails because it looks like urlopen returns bytes instead of string in that version, then I get error that write argument must be string not bytes (in `setup_convert2rhel`) 

This is the output:

```
Checking if required files are present on the system.
Checking for file /etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release
Checking for file /etc/rhsm/ca/redhat-uep.pem
Checking for file /etc/yum.repos.d/convert2rhel.repo
Downloading required files.
Writing file to destination: '/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release'
Cleaning up modifications to the system.
Removing the file '/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release' as it was previously downloaded.
File '/etc/rhsm/ca/redhat-uep.pem' was present on the system before the execution. Skipping the removal.
File '/etc/yum.repos.d/convert2rhel.repo' was present on the system before the execution. Skipping the removal.
Analysis was successfully executed.
### JSON START ###
{
    "status": "ERROR",
    "message": "",
    "report": "write() argument must be str, not bytes",
    "report_json": null
}
### JSON END ###
```

We could also make the print with `Analysis was successfully executed.` only if it actually run successfully (probably not in finaly, or change it that execution just ended?), maybe whole part that is done in finally could be visually divided from the execution prints - just to clearly see when it went bad/when it stopped

Comment (andywaltlova) [reply scripts/preconversion_assessment_script.py]: LGTM! :+1: 

Just a note: talked with @r0x0d and we decided to drop python3 for now so the error in case of wrong version is clear, output was updated as suggested in previous comment
